Add Repeater tutorial [15243] (#46)
eProsima/vulcanexus: docs/rst/tutorials/cloud/edge_edge_repeater_cloud/edge_edge_repeater_cloud.rst

intermediate NATs or dynamic networks.
In this scenario, the Cloud server will work as a **TURN** (Traversal Using Relays around NAT), a.k.a **Repeater**.
This kind of servers are meant to be accessed from every point in the WAN network, and will forward the messages from
any edge to the rest of them connected.
Thus, it creates a bridge between 2 networks that do not have direct access to each other, facilitating network configuration,
as NATs Traversals and dynamic addresses.

.. _warning_lan:

.. warning::

    This tutorial is intended for WAN communication.
    However, if there is only access to a LAN communication, it is possible to follow the tutorial by changing the ROS 2 Domain Ids so that the edges uses the different Domains each (``0`` and ``1``).
    This way the ROS 2 nodes are logically isolated and will not discovery other nodes out of their ROS 2 Domain.

This way, all the elements involved in this architecture will be studied, starting with the edge robots.
One edge will work as a controller and the other as a robot (as previously explained in :ref:`previous tutorial <tutorials_cloud_wan_edge_cloud_wan_edge_cloud>`).

> Continuing with the intermediate elements that enable communication over the Internet, in one of each edge and in Cloud.

Reviewer: How about this?

Following, all the elements involved in this architecture will be studied, starting with the edge robots and continuing with the intermediate elements that enable communication over the Internet between each edge and the Cloud.
One edge will work as a controller and the other as a robot (as previously explained in :ref:`previous tutorial <tutorials_cloud_wan_edge_cloud_wan_edge_cloud>`).

> Continuing with the intermediate elements that enable communication over the Internet, in one of each edge and in Cloud.

Reviewer: NIT:
(as already explained in previous tutorial).

The image below describes the scenario presented in this tutorial.

.. figure:: /rst/figures/tutorials/cloud/repeater_wan.png
   :align: center

Several key elements can be observed in it:

1.  **ROS 2 Application**.

>     The application used for this tutorial is the *turtlesim*.

Reviewer: Capitalize *turtlesim*.

>     The application used for this tutorial is the *turtlesim*.

Author: The name of the application is `turtlesim`. I would rather leave it as a "code" word.

>     The application used for this tutorial is the *turtlesim*.

Reviewer: I believe you can capitalize it as [here](https://docs.ros.org/en/foxy/Tutorials/Beginner-CLI-Tools/Introducing-Turtlesim/Introducing-Turtlesim.html#background), when referring to the application/simulator, and use lower case when referring to the package. It just hurts the eye like this, at least to me.

>     The *turtlesim* is a ROS 2 application, first developed for ROS, aimed at teaching the basic concepts of ROS 2 such as publish/subscribe, services and actions.

Reviewer: Capitalize.

1.  **ROS 2 Controller**.
    This is a ROS 2 application that sends commands to the edge robot.
    It has been developed a basic C++ application for this tutorial that sends publications on the topic that the ``turtlesim_node`` listens.
    By means of these publications (commands) from the controller, and the feedback information that the controller receives from the ``turtlesim_node``, it is possible to control this node automatically without the need for user intervention which facilitates the deployment of the scenario at hand.
    The key feature of the *DDS Router* is that it is easy to configure, allowing to connect different networks with ROS 2 applications without the need to apply any changes to the developer's software or applications.

1.  **ROS 2 Router / DDS Router**.
    *eProsima ROS 2 Router*, a.k.a `DDS Router <https://github.com/eProsima/DDS-Router>`_, is an end-user software application that enables the connection of distributed ROS 2 networks (see DDS Router documentation `here <https://eprosima-dds-router.readthedocs.io/en/latest/>`).

>     That is, ROS 2 nodes such as publishers and subscriptions, or clients and services, deployed in one geographic location and using a dedicated local network will be able to communicate with other ROS 2 nodes deployed in different geographic areas on their own dedicated local networks as if they were all on the same network through the use of *DDS Router*.

Author: For unknown reasons, in ros2 code is called subscription, but in ros2 tutorials is referenced as subscriber. Which one do you prefer?

>     That is, ROS 2 nodes such as publishers and subscriptions, or clients and services, deployed in one geographic location and using a dedicated local network will be able to communicate with other ROS 2 nodes deployed in different geographic areas on their own dedicated local networks as if they were all on the same network through the use of *DDS Router*.

Reviewer: Your choice, I'm more familiarized with subscriber but as you wish.

    * *DDS Router Repeater*. It plays the TURN server role in the communication. It will expose some public network addresses to which the edge Routers could establish communication, and will forward the message from one LAN to the other.

Prerequisites
-------------

This tutorial will require two machines (*Robot 1* and *Cloud Server*) deployed on different networks (*LAN 1* and *Cloud*).
It is possible to simulate the scenario by deploying everything needed on the same machine and two virtual networks but let's focus on the case of a real deployment.

It is also important to have previously installed Vulcanexus using one of the following installation methods:

* :ref:`linux_binary_installation`
* :ref:`linux_source_installation`
* :ref:`docker_installation`

Deployment on LAN 1
-------------------

First, let's deploy the ``turtlesim_node`` and DDS Router Edge on a machine on *LAN 1*.

Running turtlesim_node on the edge
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

First, setup the Vulcanexus environment to have the ``turtlesim_node`` available.
For this there are two options:

> 1.  Running the Vulcanexus Docker image.

Reviewer: Still wrong, I believe it is the blank line.

Running DDS Router Edge 1
^^^^^^^^^^^^^^^^^^^^^^^^^

Then, to run the DDS Router Edge configure the environment as in the previous step.

.. note::

    If you are deploying Vulcanexus from the Docker image, note that you will need to have a configuration file (``config.yaml``) to configure the DDS Router Edge accessible from your Docker container.

    You can do this by mounting a shared volume when launching the container, by copying the file from the local host to the container in case it is already running, or by editing a file from the Docker container itself.

Setup the Vulcanexus environment, either in a Docker container or on the local host, running the following command:

.. code-block:: bash

    source /opt/vulcanexus/humble/setup.bash

Let's create the DDS Router configuration file. It will look like the one shown below.

.. literalinclude:: /resources/tutorials/cloud/edge_edge_repeater_cloud/dds_router_edge_1.yaml
    :language: yaml

Next, it is briefly explained the most relevant aspects of this configuration file.

The ``participants`` are the interfaces of the DDS Router to communicate with other networks. In this case, we have two participants:

>     *   ``local``: this is a simple participant that communicates with all ROS 2 nodes it finds in domain 1.

Reviewer: The turtlesim_node is running in domain `1` though. You should change that and instead run the controller in this domain.

>     *   ``local``: this is a simple participant that communicates with all ROS 2 nodes it finds in domain 1.

Reviewer: Almost there, the controller still needs to run in domain `1`.

>     *   ``local``: this is a simple participant that communicates with all ROS 2 nodes it finds in domain 1.

Author: It render incorrectly on local

>     *   ``local``: this is a simple participant that communicates with all ROS 2 nodes it finds in domain 1.

Reviewer: What is missing is the WAN/LAN tabs for the router edge 2 yaml file.

    *   ``repeater-client``: it is a participant designed to communicate with a Repeater Participant.
        It uses the `Fast DDS Initial Peers discovery mechanism <https://fast-dds.docs.eprosima.com/en/latest/fastdds/discovery/simple.html#initial-peers>`_ to establish a point-to-point communication between two DDS entities, two *DDS Routers* in this case.

    For the DDS Router Edge, a connection address shall be defined which must be the same as the one exposed by the Cloud Server.
    There are some relevant configurations within this connection address:

    * ``connection-addresses``: defines the IP (``ip``) and port (``port``) of the network addresses to which it connects, and the transport protocol (``transport``) to be used in the communication, TCP in this case.

.. note::

    In this case, the DDS Router will forward all topics found in the network.
    However, it is important to mention that the ROS 2 topics relayed by the DDS Router can be filtered by configuring the ``allowlist`` and ``blocklist``.

>     If this is the case please refer to the `DDS Router documentation <https://eprosima-dds-router.readthedocs.io/en/latest/>`_ for information on how to do this.

Reviewer: Maybe better share the corresponding configuration section directly.

The following figure summarizes the deployment on the edge.

.. todo::

    add new image

> .. figure:: /rst/figures/tutorials/cloud/edge_1_repeater_deployment.png

Reviewer: Again, make sure the same domain is being used.

Author: I do not understand the suggestion. In the image domain 0 is used

Reviewer: See above.

   :align: center

To finish this step, run the DDS Router with the configuration file created as an argument.

.. code-block::

    ddsrouter -c <path/to/file>/ddsrouter_edge_1.yaml

Deployment of LAN 2
-------------------

> Running the turtlesim_square_move on the Cloud

Reviewer: on the cloud?

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. todo::

>     decide wether to link to previous tutorial or reexplain it here

Reviewer: For completion, I would explain here as well.

    .. tab:: LAN

        .. code-block:: bash

            ROS_DOMAIN_ID=1 ros2 run turtlesim turtlesim_square_move

        .. note::

            As stated :ref:`here <warning_lan>`, change the ROS 2 Domain Id if running the edge and cloud applications on the same LAN.

The important points to note in this application are the following:

*   The control application sends the movement commands to the ``turtlesim_node`` through a publisher in the `geometry_msgs/msg/Twist <https://docs.ros2.org/foxy/api/geometry_msgs/msg/Twist.html>`_ topic.
    This topic expresses the velocity at which the turtle has to move divided into linear velocity (``linear``) and angular velocity (``angular``).

*   The application knows the position of the turtle on the map at any moment and whether it is stopped or moving by subscribing to the `turtlesim/msg/pose <https://docs.ros2.org/foxy/api/turtlesim/msg/Pose.html>`_ topic.
    This topic provides information about the turtle's coordinates (``x`` and ``y``) and the turtle's rotation (``theta``). We can also know its linear and angular velocity (``linear_velocity`` and ``angular_velocity``).

Running DDS Router Edge 2
^^^^^^^^^^^^^^^^^^^^^^^^^

> As the Repeater server is the same for both edges, the configuration of the DDS Router Edge 2 is similar same as

Reviewer: ```suggestion
As the Repeater server is the same for both edges, the configuration of the DDS Router Edge 2 is very similar to 
```

:ref:`the one for the DDS Router Edge 1 <tutorials_cloud_edge_edge_repeater_cloud_edge_edge_repeater_cloud_running_router_1>`.
In this example both edges use different ports to communicate with the Repeater, simulating 2 different networks available in the Cloud.
However this is not needed, and only one address could be used.
The following snippet shows a configuration file (changing Domain for LAN scenarios):

.. literalinclude:: /resources/tutorials/cloud/edge_edge_repeater_cloud/dds_router_edge_2.yaml
    :language: yaml

The following figure summarizes the deployment on the edge 2.

.. todo::

    add new image

.. figure:: /rst/figures/tutorials/cloud/edge_2_repeater_deployment.png
   :align: center

> To finish this step, run the DDS Router with the configuration file created as an argument.

Reviewer: ```suggestion
Now, run the DDS Router with the configuration file created as an argument.
```

.. code-block::

    ddsrouter -c <path/to/file>/ddsrouter_edge_2.yaml

Deployment of Cloud Repeater
----------------------------

> In order to communicate both edges, we use a DDS Router configuring as a *Repeater* that forward the messages from one edge to the other.

Reviewer: ```suggestion
In order to communicate both edges, a DDS Router configured as *Repeater* is used, forwarding the messages from one edge to the other.
```

This machine should be accessible from the Internet.

> In case this device is working under a NAT, check :ref:`previous turotial <tutorials_cloud_wan_edge_cloud_wan_edge_cloud_configure_transversal_nat>`

Reviewer: ```suggestion
In case this device is working under a NAT, check :ref:`previous tutorial <tutorials_cloud_wan_edge_cloud_wan_edge_cloud_configure_transversal_nat>`
```

for more information about how to configure the NAT to be accessible from the outside.

> The following snippet shows a configuration file for this DDS Router:

Reviewer: ```suggestion
The following snippet shows the configuration file for this DDS Router:
```

.. literalinclude:: /resources/tutorials/cloud/edge_edge_repeater_cloud/dds_router_repeater.yaml
    :language: yaml

In this case, there is only one Participant configured as ``repeater``.
This Participant will wait for external Participants to communicate via Initial Peers from its ``listening-addresses``.

> Once the discovery has been achieved, this Repeater will forward the data from one edge to the other.

Reviewer: ```suggestion
Once the discovery of clients occurs, this Repeater will forward the data from one edge to the other.
```

> To run the DDS Router with the configuration file created as an argument, execute the following command after sourcing the Vulcanexus environment:

Reviewer: ```suggestion
To run the DDS Router with the configuration file created as an argument, execute the following command after having sourced the Vulcanexus environment:
```
